### PR TITLE
fix(apigateway): base path url cannot contain upper case characters

### PR DESCRIPTION
**Pull request by nicklaw5**
We're migrating an existing API Gateway application to CDK and have [hit this error](https://github.com/aws/aws-cdk/blob/2f0a598720f7ecf4dfea7dde97182940fcbe8d9a/packages/%40aws-cdk/aws-apigateway/lib/base-path-mapping.ts#L52), because our base paths include capitalised characters.

This PR resolves that issue and I've included some basics tests for the `BasePathMapping` resource.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nicklaw5** [comment]
Hi @nija-at. Do you know when this PR will be reviewed? This bug is blocker on our end.

File: packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    });
+
+    // THEN
+    expect(stack).toHaveResource('AWS::ApiGateway::BasePathMapping', {
+      DomainName: { Ref: 'MyDomainE4943FBC' },
+      RestApiId: { Ref: 'MyApi49610EDF' },
+    });
+  });
+
+  test('specify basePath property', () => {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts:31]: Thanks for adding these tests. 

Can you also add a test case where basepath has an illegal character and assert that the expected error is produced?

**Review by nija-at — DISMISSED** [review]
Nice! Thanks for submitting this PR. Just one comment below.

File: packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts (hunk 1 above)
Comment (nicklaw5) [reply packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts:31]: @nija-at  I've added the requested test case.

File: packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    const invalidBasePath = '/invalid-/base-path';
+
+    // THEN
+    expect(() => {
+      new apigw.BasePathMapping(stack, 'MyBasePath', {
+        restApi: api,
+        domainName: domain,
+        basePath: invalidBasePath,
+      });
+    }).toThrowError(`A base path may only contain letters, numbers, and one of "$-_.+!*'()", received: ${invalidBasePath}`);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts:76]:
```suggestion
    }).toThrowError(/base path may only only contain/);
```

**Review by nija-at — APPROVED** [review]

File: packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    const invalidBasePath = '/invalid-/base-path';
+
+    // THEN
+    expect(() => {
+      new apigw.BasePathMapping(stack, 'MyBasePath', {
+        restApi: api,
+        domainName: domain,
+        basePath: invalidBasePath,
+      });
+    }).toThrowError(/base path may only only contain/);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/test/base-path-mapping.test.ts:76]:
```suggestion
    }).toThrowError(/base path may only contain/);
```

**Review by nija-at — APPROVED** [review]

**Pull request merged**